Enhancement/facebook publicize error

cerezo074 opened the pull request:
Fixes #12228 just the ios application. **1 copy is missing I couldn't find the translation.** See changes over localizable file.

<img src="https://user-images.githubusercontent.com/6471815/63241404-30310280-c219-11e9-9bf8-7fc868b31aeb.png" width="320">

![Simulator Screen Shot - iPad Air - 2019-08-18 at 23 38 42 copia](https://user-images.githubusercontent.com/6471815/63241387-260f0400-c219-11e9-80c2-d1d2f44fda00.png)

To test:
Just share over Facebook with an account that doesn't have a Facebook page.
Update release notes:

- [x] If there are user-facing changes, I have added an item to `RELEASE-NOTES.txt`.
UX and UI is impacted a little by this enhancement, so i've added a note inside the document explaining what is the change. This is my first PR let me know if i did some mistake, i will fix it.

13.1 - **Shows an alert instead of showing a new screen for facebook publicize error.**

## Tasks 

- [ ] Create a separate bug report for https://github.com/wordpress-mobile/WordPress-iOS/pull/12337#issuecomment-526018568
- [x] Revert Localizable.strings changes https://github.com/wordpress-mobile/WordPress-iOS/pull/12337#issuecomment-526674001

shiki commented:
@cerezo074 Thank you for working on this. I can review this but wanted to confirm first if this is ready for review. Are you still working on the responses at https://github.com/wordpress-mobile/WordPress-Android/issues/10278#issuecomment-522353053?

cerezo074 commented:
@shiki I did a simple change requested by @mattmiklic, and I used the interface for using localizables inside the project **NSLocalizedString(@"localizable key", @"comment")** I think it was done as you suggested, but if look the localizable files modified by this pr it was just 1(ENG). Dont we need to insert the new key for the other languanges? that was my question.

shiki commented on WordPress/Classes/ViewRelated/Blog/KeyringAccountController.swift:
```diff
@@ -0,0 +1,53 @@
+    var keyringConnection: KeyringConnection
+}
+
+@objc class KeyringAccountController: NSObject {
```
Would you mind explaining why this class had to exist? I'm also unsure of `*Controller` suffix here. That usually denotes a `ViewController` subclass but this class is not.

cerezo074 replied:
This is a helper class, it encapsulates logic for KeringAccount. I've seen that you don't have a ViewModel or a Presenter so i don't think that UI(include view controller) needs to know about this logic. Another reason is that i want to rehuse code and avoid duplicating that. Let me change that to KeyringAccountHelper, what do you think about it?

shiki commented on RELEASE-NOTES.txt:
```diff
@@ -8,7 +8,8 @@
 * Stats: added File Downloads to period stats.
 * Adds a Publish Now action to posts in the posts list. 
- 
+* Shows an alert instead of showing a new screen for facebook publicize error.
```
This should be in the 13.2 now.

cerezo074 commented:
Hi @shiki, thanks for the comments, i sent new changes regarding to your comments, this time i've made alert information more generic, the idea was to avoid SharingAuthorizationHelper knows about facebook stuff and just let another class takes the responsability(KeyringAccountHelper) to handle it and create the information depending on which account type was used.

cerezo074 commented:
@shiki, i fixed that mistake but there is something that catch my attention am not sure if this bug is being tracked down. But for some reason when i used an account with a valid page, and after i switched to another account that haven't had any pages(i can see the button for creating a new one), result when i finished the facebook flow and returned to the app is that i am seeing the page for the previous account. Steps that i did:

1. log in with an account that have a valid page
2. finish the flow an return the app, you can see the page listed.
3. close the screen(select the "X")
4. log in with an account that doesn't have any page(validate that you can't see any page and you can create one)
5. Finish the flow, after you return to the app, you can see the page listed

I test over bb8c73ce98953a38b7061f30a2f21de78d7c68d9 commit from your develop branch this behaviour is happening with my branch too.

shiki commented:
> I test over bb8c73c commit from your develop branch this behaviour is happening with my branch too. 

@cerezo074 That does sound like a different problem. Would you mind creating a separate issue for it?

cerezo074 commented:
@shiki, changes were made it as you suggested, let me know if everything is as you expected 🤟.

shiki commented on WordPress/Classes/ViewRelated/Blog/SharingAuthorizationHelper.m:
```diff
@@ -231,32 +231,25 @@ - (void)fetchKeyringConnectionsForService:(PublicizeService *)pubServ
 }
 
--(void)showAlertFromConnectionsWithKeyringAccountHelper:(KeyringAccountHelper *)keyringAccountHelper
+-(void)showAlertFromConnectionsWithValidationError:(ValidationError *)validationError
```
Since this is only called once, do you think it's a good idea to just get rid of this and go straight with calling

```objc
    [weakSelf.delegate sharingAuthorizationHelper:self
                 requestToShowValidationError:validationError
                           fromViewController:self.navController];
```

... in `fetchKeyringConnectionsForService` above?

shiki commented on WordPress/Classes/ViewRelated/Blog/KeyringAccountHelper.swift:
```diff
@@ -0,0 +1,101 @@
+    func facebookPublicizeAlertFields() -> ConfirmationAlertFields {
+        let alertHeaderMessage = NSLocalizedString("Not connected",
+                                                   comment: "Error title for alert, shown to a user who is trying to share to Facebook but does not have any available Facebook Pages.")
+        let alertBodyMessage = NSLocalizedString("The Facebook connection could not be made because this account does not have access to any pages. Facebook supports sharing connections to Facebook Pages, but not to Facebook Profiles.",
```
@cerezo074 It doesn't look like this has been fixed yet.

shiki commented on WordPress/Classes/ViewRelated/Blog/SharingConnectionsViewController.m:
```diff
@@ -231,6 +231,15 @@ - (void)handleConnectTapped:(NSIndexPath *)indexPath
     [self.helper connectPublicizeService];
 }
 
+- (void)handleLearnMoreTappedWithLearnMoreURL:(NSURL*)learnMoreURL
```
I think it makes sense that you stuck with `ValidationError.continueURL` instead of my suggested `ValidationError.learnMoreURL`. However, I think we should be consistent. Since you stuck with `continueURL`, we should call this:

```suggestion
- (void)handleContinueURLTapped:(NSURL*)url
```

cerezo074 commented:
@shiki I did the change of the alert body, this time I modified the localizable resources to match the new key and in some languages that have the same value, I changed them too. Other that have different translations I couldn't change it but I think that you have a team that do those changes. Let me know if you need more changes. 😅

shiki commented:
@cerezo074 

> this time I modified the localizable resources to match the new key and in some languages that have the same value, I changed them too.

Oh, no. Please don't change the `Localizable.strings`. Those are automatically generated and as I mentioned in https://github.com/wordpress-mobile/WordPress-Android/issues/10278#issuecomment-522764410, you shouldn't need to worry about them as long as you use `NSLocalizedString()` in the code. 

Would you mind reverting those changes?

cerezo074 commented:
That's easy, just for curiosity, how do you do that automatically?.

shiki commented on WordPress/Classes/ViewRelated/Blog/KeyringAccountHelper.swift:
```diff
@@ -0,0 +1,99 @@
+private extension KeyringAccountHelper {
+
+    func facebookPublicizeAlertFields() -> ValidationError {
+        let alertHeaderMessage = NSLocalizedString("Not connected",
```
Sorry I didn't notice this before, the “connected“ should be capitalized based on the screenshot in https://github.com/wordpress-mobile/WordPress-Android/issues/10278#issuecomment-522613358:

<img src="https://user-images.githubusercontent.com/1396552/63275712-035c0a00-c270-11e9-9156-cd03c616527b.png">

cerezo074 commented:
@shiki, what should i do? there are some conflicts due to some commits that @jkmassel did it for 13.2, if you want i can revert those commit and my merge with the release/13.2 branch. Let me know if that works for you.

shiki commented:
@cerezo074 Merged! Thank you very much!
